crates/services/p2p/src/codecs/postcard.rs

```diff
@@ -1,219 +1,85 @@
-    where
-        T: AsyncRead + Unpin + Send,
-    {
-        let mut response = Vec::new();
-        socket
-            .take(self.max_response_size as u64)
-            .read_to_end(&mut response)
-            .await?;
-        deserialize(&response)
-    }
-
-    async fn read_response<T>(
-        &mut self,
-        protocol: &Self::Protocol,
-        socket: &mut T,
-    ) -> io::Result<Self::Response>
-    where
-        T: AsyncRead + Unpin + Send,
-    {
-        let mut response = Vec::new();
-        socket
-            .take(self.max_response_size as u64)
-            .read_to_end(&mut response)
-            .await?;
-
-        match protocol {
-            PostcardProtocol::V1 => {
-                let v1_response = deserialize::<V1ResponseMessage>(&response)?;
-                Ok(v1_response.into())
-            }
-            PostcardProtocol::V2 => deserialize::<V2ResponseMessage>(&response),
+impl GossipsubMessageHandler<PostcardCodec> {
+    pub fn new() -> Self {
+        GossipsubMessageHandler {
+            codec: PostcardCodec,
         }
     }
-
-    async fn write_request<T>(
-        &mut self,
-        _protocol: &Self::Protocol,
-        socket: &mut T,
-        req: Self::Request,
-    ) -> io::Result<()>
-    where
-        T: futures::AsyncWrite + Unpin + Send,
-    {
-        let encoded_data = serialize(&req)?;
-        socket.write_all(&encoded_data).await?;
-        Ok(())
-    }
-
-    async fn write_response<T>(
-        &mut self,
-        protocol: &Self::Protocol,
-        socket: &mut T,
-        res: Self::Response,
-    ) -> io::Result<()>
-    where
-        T: futures::AsyncWrite + Unpin + Send,
-    {
-        let encoded_data = match protocol {
-            PostcardProtocol::V1 => {
-                let v1_response: V1ResponseMessage = res.into();
-                serialize(&v1_response)?
-            }
-            PostcardProtocol::V2 => serialize(&res)?,
-        };
-        socket.write_all(&encoded_data).await?;
-        Ok(())
-    }
 }
 
-impl GossipsubCodec for PostcardCodec {
-    type RequestMessage = GossipsubBroadcastRequest;
-    type ResponseMessage = GossipsubMessage;
-
-    fn encode(&self, data: Self::RequestMessage) -> Result<Vec<u8>, io::Error> {
-        let encoded_data = match data {
-            GossipsubBroadcastRequest::NewTx(tx) => postcard::to_stdvec(&*tx),
-        };
-
-        encoded_data.map_err(|e| io::Error::new(io::ErrorKind::Other, e.to_string()))
-    }
-
-    fn decode(
-        &self,
-        encoded_data: &[u8],
-        gossipsub_tag: GossipTopicTag,
-    ) -> Result<Self::ResponseMessage, io::Error> {
-        let decoded_response = match gossipsub_tag {
-            GossipTopicTag::NewTx => GossipsubMessage::NewTx(deserialize(encoded_data)?),
-        };
-
-        Ok(decoded_response)
+impl<T> Encode<T> for PostcardCodec
+where
+    T: ?Sized + serde::Serialize,
+{
+    type Encoder<'a> = Cow<'a, [u8]> where T: 'a;
+    type Error = io::Error;
```

Comment:
Why do we return an IO error here instead of just propagating the postcard error?

Reply:
Previously the implementation of serialize was 

```rust 
fn serialize<D: Serialize>(data: &D) -> Result<Vec<u8>, io::Error> {
    postcard::to_stdvec(&data)
        .map_err(|e| io::Error::new(io::ErrorKind::Other, e.to_string()))
}
```
so I left the same error type. 
We can change the error type, but I would not do it in this PR. I usually avoid introducing any non-functional changes in refactoring PRs (that also includes fixing bugs), as any functional change should be handled separately (or at least in a separate commit).

Reply:
Fair enough :+1: